File: Web4-job-runner/Main.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"math/rand"
	"net/http"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"github.com/google/uuid"
)

// --------------------- CONFIG ---------------------
type Config struct {
	MaxConcurrency int        `json:"max_concurrency"`
	MaxRetries     int        `json:"max_retries"`
	Tasks          []TaskSpec `json:"tasks"`
}

type TaskSpec struct {
	Type    string `json:"type"`
	Payload string `json:"payload"`
}

func loadConfig() Config {
	cfg := Config{
		MaxConcurrency: 3,
		MaxRetries:     2,
		Tasks: []TaskSpec{
			{Type: "download", Payload: "https://httpbin.org/get"},
			{Type: "ai", Payload: "Write Web4 article summary"},
			{Type: "blockchain", Payload: "0xContractAddress:mintNFT"},
			{Type: "storage", Payload: "/tmp/sample.txt"},
		},
	}

	if v := os.Getenv("CONFIG_JSON"); v != "" {
		json.Unmarshal([]byte(v), &cfg)
	}
	return cfg
}

// --------------------- TASK ---------------------
type Task struct {
	ID         string // UUID for global uniqueness
	Spec       TaskSpec
	MaxRetries int
}

func (t Task) Run(ctx context.Context) {
	for attempt := 0; attempt <= t.MaxRetries; attempt++ {
		select {
		case <-ctx.Done():
			logTask(t.ID, attempt, "Task canceled")
			return
		default:
		}

		start := time.Now()
		logTask(t.ID, attempt, fmt.Sprintf("Starting task type=%s payload=%s", t.Spec.Type, t.Spec.Payload))
		err := t.execute(ctx)
		duration := time.Since(start).Seconds()

		if err != nil {
			logTask(t.ID, attempt, fmt.Sprintf("Attempt %d failed after %.2fs: %v", attempt, duration, err))
			if attempt < t.MaxRetries {
				backoff := time.Duration(500*int64(1<<attempt)) * time.Millisecond
				time.Sleep(backoff)
			}
			continue
		}

		logTask(t.ID, attempt, fmt.Sprintf("Attempt %d succeeded in %.2fs", attempt, duration))
		break
	}
}

func (t Task) execute(ctx context.Context) error {
	switch t.Spec.Type {
	case "download":
		return taskDownload(ctx, t.Spec.Payload, t.ID)
	case "ai":
		return taskAI(ctx, t.Spec.Payload, t.ID)
	case "blockchain":
		return taskBlockchain(ctx, t.Spec.Payload, t.ID)
	case "storage":
		return taskStorage(ctx, t.Spec.Payload, t.ID)
	default:
		return fmt.Errorf("unknown task type %s", t.Spec.Type)
	}
}

// --------------------- TASK TYPES ---------------------
func taskDownload(ctx context.Context, url, taskID string) error {
	req, _ := http.NewRequestWithContext(ctx, "GET", url, nil)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("download failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return fmt.Errorf("download failed with status %d", resp.StatusCode)
	}

	// Use task UUID for unique filename
	fileName := fmt.Sprintf("task_download_%s.html", taskID)
	out, err := os.Create(fileName)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, resp.Body)
	return err
}

func taskAI(ctx context.Context, prompt, taskID string) error {
	select {
	case <-ctx.Done():
		return fmt.Errorf("AI task canceled")
	case <-time.After(500 * time.Millisecond):
	}
	log.Printf("[AI %s] Generated content for prompt: %s", taskID, prompt)
	return nil
}

func taskBlockchain(ctx context.Context, action, taskID string) error {
	select {
	case <-ctx.Done():
		return fmt.Errorf("Blockchain task canceled")
	case <-time.After(300 * time.Millisecond):
	}
	if rand.Float64() < 0.2 {
		return fmt.Errorf("blockchain tx failed")
	}
	log.Printf("[Blockchain %s] Executed action: %s", taskID, action)
	return nil
}

func taskStorage(ctx context.Context, path, taskID string) error {
	select {
	case <-ctx.Done():
		return fmt.Errorf("Storage task canceled")
	case <-time.After(200 * time.Millisecond):
	}
	log.Printf("[Storage %s] Uploaded file: %s", taskID, path)
	return nil
}

// --------------------- LOGGING ---------------------
func logTask(taskID string, attempt int, msg string) {
	entry := map[string]interface{}{
		"timestamp": time.Now().Format(time.RFC3339),
		"taskID":    taskID,
		"attempt":   attempt,
		"message":   msg,
	}
	data, _ := json.Marshal(entry)
	log.Println(string(data))
}

// --------------------- MAIN ---------------------
func main() {
	rand.Seed(time.Now().UnixNano())
	cfg := loadConfig()
	log.Printf("Web4 Job Runner Configuration: %+v", cfg)

	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer cancel()

	var wg sync.WaitGroup
	sem := make(chan struct{}, cfg.MaxConcurrency)

	for _, spec := range cfg.Tasks {
		wg.Add(1)
		sem <- struct{}{}

		// Generate a new UUID per task
		taskID := uuid.New().String()

		go func(ts TaskSpec, tID string) {
			defer wg.Done()
			Task{ID: tID, Spec: ts, MaxRetries: cfg.MaxRetries}.Run(ctx)
			<-sem
		}(spec, taskID)
	}

	wg.Wait()
	log.Println("Web4 Job Runner complete!")
}

File: Web4-job-runner/Main.dart
import 'dart:io';
import 'dart:math';
import 'dart:async';
import 'package:uuid/uuid.dart';

var uuid = Uuid();
var random = Random();

void main() async {
  // Example tasks
  var tasks = [
    {"type": "download", "payload": "https://httpbin.org/get"},
    {"type": "ai", "payload": "Write a Web4 article summary"},
    {"type": "blockchain", "payload": "mintNFT"},
    {"type": "storage", "payload": "/tmp/sample.txt"},
    {"type": "download", "payload": "https://httpbin.org/anything"},
  ];

  int maxConcurrency = 3;
  var semaphore = Semaphore(maxConcurrency);

  // Run all tasks concurrently with maxConcurrency limit
  var futures = tasks.map((task) async {
    await semaphore.acquire();
    try {
      var taskId = uuid.v4();
      await runTask(taskId, task);
    } finally {
      semaphore.release();
    }
  }).toList();

  await Future.wait(futures);
  print("All tasks complete!");
}

// ----------------- Concurrency control -----------------
class Semaphore {
  int _maxConcurrent;
  int _current = 0;
  final _queue = <Completer>[];

  Semaphore(this._maxConcurrent);

  Future<void> acquire() {
    if (_current < _maxConcurrent) {
      _current++;
      return Future.value();
    }
    var completer = Completer<void>();
    _queue.add(completer);
    return completer.future;
  }

  void release() {
    if (_queue.isNotEmpty) {
      _queue.removeAt(0).complete();
    } else {
      _current--;
    }
  }
}

// ----------------- Task runner -----------------
Future<void> runTask(String taskId, Map task) async {
  print("[$taskId] Starting ${task['type']} task...");

  switch (task['type']) {
    case "download":
      await taskDownload(taskId, task['payload']);
      break;
    case "ai":
      await taskAI(taskId, task['payload']);
      break;
    case "blockchain":
      await taskBlockchain(taskId, task['payload']);
      break;
    case "storage":
      await taskStorage(taskId, task['payload']);
      break;
    default:
      print("[$taskId] Unknown task type: ${task['type']}");
  }

  print("[$taskId] Task complete!\n");
}

// ----------------- Task implementations -----------------
Future<void> taskDownload(String taskId, String url) async {
  await Future.delayed(Duration(milliseconds: 500 + random.nextInt(500)));
  var fileName = "download_$taskId.html";
  File(fileName).writeAsStringSync("<html><body>Downloaded from $url</body></html>");
  print("[$taskId] Download saved as $fileName");
}

Future<void> taskAI(String taskId, String prompt) async {
  await Future.delayed(Duration(milliseconds: 500));
  print("[$taskId] AI generated content for prompt: $prompt");
}

Future<void> taskBlockchain(String taskId, String action) async {
  await Future.delayed(Duration(milliseconds: 300));
  if (random.nextDouble() < 0.2) {
    print("[$taskId] Blockchain action $action failed!");
    return;
  }
  print("[$taskId] Blockchain action $action succeeded!");
}

Future<void> taskStorage(String taskId, String path) async {
  await Future.delayed(Duration(milliseconds: 200));
  print("[$taskId] File $path uploaded to storage");
}
